src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+        this instance.
+
+        NOTE: the necessary dimensions for the jacobian can be extracted from the dimensions
+        of the jacobian of passed Ad_array instances.
+        """
+        pass
+
+
+class AbstractJacFunction(AbstractFunction):
+    """'Half'-abstract base class, for which a direct evaluation of the passed callable is
+    implemented. Only the Jacobian has to be concretized / approximated."""
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        """
+        Evaluates the passed callable directly using values of Ad_array instances
+        passed as arguments.
+
+        :param args: tuple of :class:`~porepy.numerics.ad.forward_mode.Ad_array`
+        :type args: tuple
+        :return: results from callable packed into an array.
+        :rtype: numpy.array
+        """
+        # get values of argument Ad_arrays.
+        vals = (arg.val for arg in args)
+
+        # if the callable is flagged as conform for vector operations, feed vectors
+        if self.is_vector_func:
+
+            return self.func(*vals)
+        else:
+            # if not vector-conform, feed element-wise
+
+            # TODO this displays some special behavior when val-arrays have different lengths:
+            # it returns None-like things for every iteration more then shortest length
+            # These Nones are ignored for some reason by the function call, as well as by the
+            # array constructor.
+            # If a mortar var and a subdomain var are given as args,
+            # then the lengths will be different for example.
+            return np.array([self.func(*vals_i) for vals_i in zip(*vals)])
+
+
+### CONCRETE IMPLEMENTATIONS ------------------------------------------------------------------
+
+
+class Function(AbstractFunction):
+    """Ad representation of an analytically given function,
+    where it is expected that feeding the callable with Ad_arrays directly will give
+    give the proper result.
+
+    Here the values AND the Jacobian are obtained exactly by standard rules of differentiation.
+
+    The intended use is as a wrapper for operations on pp.ad.Ad_array objects,
+    in forms which are not directly or easily expressed by the rest of the Ad
+    framework.
+
+    NOTE: this is a special case where `get_values` and `get_jacobian` are formally implemented
+    but never used by the AD framework
+    (due to the assumed, direct evaluation of result using the callable).
+    """
+
+    def __init__(
+        self, func: Callable, name: str, is_vector_func: Optional[bool] = True
+    ):
+        super().__init__(func, name, is_vector_func, True)
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.val
+
+    def get_jacobian(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.jac
+
+
+class LJacFunction(AbstractJacFunction):
```

Comment:
The naming here is misleading (the L-scheme is just a special case of quasi-Newton method with a fixed Jacobian with a certain structure, it has nothing to do with approximation of derivatives per se), it should be something like DiagonalJacFunction. More important, I would rather have a function with a preset sparse matrix as the Jacobian and value. Although assigning diagonal matrices may be an important special case, it is not at all clear this deserves a special class.

I have the feeling the superclasses may be intended to cover the case of more general approximated Jacobian matrices - if so, see comments elsewhere about naming conventions and the class hierarchy.

Reply:
This class is a first working example. I agree that it's wider usability is not given.